src/service/loading-indicator.js

```diff
@@ -0,0 +1,179 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {createLoaderElement} from './loader-element';
+import {htmlFor} from '../static-template';
+import {isIframed, removeElement} from '../dom';
+import {registerServiceBuilderForDoc} from '../service';
+
+const MIN_SIZE = 20;
+
+/**
+ * @typedef {{
+ *   shown: boolean,
+ *   loader: !Element,
+ *   container: !Element,
+ * }}
+ */
+let LoadingIndicatorStateDef;
+
+/**
+ * @param {!Node|!./ampdoc-impl.AmpDoc} nodeOrDoc
+ */
+export function installLoadingIndicatorForDoc(nodeOrDoc) {
+  registerServiceBuilderForDoc(
+    nodeOrDoc,
+    'loadingIndicator',
+    LoadingIndicatorImpl
+  );
+}
+
+/**
+ * @implements {../service.Disposable}
+ */
+export class LoadingIndicatorImpl {
+  /**
+   * @param {!./ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    /** @private @const */
+    this.ampdoc_ = ampdoc;
+
+    const {win} = ampdoc;
+    const inViewport = this.inViewport_.bind(this);
+    const ioCallback = (records) =>
+      /** @type {!Array<!IntersectionObserverEntry>} */ (records).forEach(
+        inViewport
+      );
+    const iframed = isIframed(win);
+    const rootMargin = '25%';
+    // Classic IntersectionObserver doesn't support viewport tracking and
```

Comment:
What do you think is a reasonable plan for improving this situation? When the polyfill supports rootMargin in x-origni iframes, do you think its reasonable to use that instead of the 3 situations we have today

Reply:
I think we need to delegate this decision to the full loader "intersect-resources" launch/cleanup. If we lose 25% margin for inViewport in some cases is a lot less critical vs losing loader functionality. So, in that scope we will have to decide what we do. It's very possible that we may chose to:
1. Add `root=document` feature to the polyfill. This would be a relatively straightforward thing to do. It's actually mostly done already.
2. Override of the native InOb that doesn't support `root=document` with the polyfill. This is, of course, a lot more controversial. For now this is a majority case, though I think it should be reducing quickly. But, certainly, using the polyfill where the native is available is super inefficient.

To remedy this, we could also be very selective when we override. E.g. we can override _only_ when we are iframed and _only_ when `root=document` is requested. E.g.

```
class IntersectionObserverSelector {
  constructor(callback, options) {
    const constr =
      isIframed() && options.root ?
      IntersectionObserverPolyfill :
      IntersectionObserverNative;
    return new constr(callback, options);
  }
}
```

But, this kind of effort is only worth it for loader, not inViewport.